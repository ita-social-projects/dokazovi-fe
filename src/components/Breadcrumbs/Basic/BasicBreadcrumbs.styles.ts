import { makeStyles } from '@material-ui/core';

export const useStyles = makeStyles(
  {
    breadcrumbs: {
      fontFamily: 'Literata',
      fontStyle: 'italic',
      fontWeight: 'normal',
      fontSize: '11px',
      lineHeight: '28px',
      color: '#2F80ED',
      position: 'absolute',
      right: '30px',
      top: '20px',
    },
    link: { textDecoration: 'underline' },
    postTitle: {
      color: '#767676',
      fontSize: 'inherit',
    },
  },
  { name: 'BasicBreadcrumbs' },
);
