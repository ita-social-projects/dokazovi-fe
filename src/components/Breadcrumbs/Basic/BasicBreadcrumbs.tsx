import { Breadcrumbs, Typography } from '@material-ui/core';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';
import { langTokens } from '../../../locales/localizationInit';
import { useStyles } from './BasicBreadcrumbs.styles';

export type BreadcrumbsLinksType = { to: string; label: string }[];

interface IProps {
  breadcrumbsLinks: BreadcrumbsLinksType;
  title: string;
}

export const BasicBreadcrumbs: React.FC<IProps> = ({
  breadcrumbsLinks,
  title,
}) => {
  const { t } = useTranslation();
  const classes = useStyles();

  return (
    <Breadcrumbs aria-label="breadcrumb" className={classes.breadcrumbs}>
      <Link className={classes.link} to="/">
        {t(langTokens.common.main)}
      </Link>
      {breadcrumbsLinks.map(({ to, label }) => (
        <Link data-testid="link" className={classes.link} key={to} to={to}>
          {label}
        </Link>
      ))}
      <Typography className={classes.postTitle}>{title}</Typography>
    </Breadcrumbs>
  );
};
