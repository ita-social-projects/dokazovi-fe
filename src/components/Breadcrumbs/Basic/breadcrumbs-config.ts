import type { BreadcrumbsContextType } from './BreadcrumbsProvider';

export enum PostBreadcrumbsState {
  fromMaterialsPage = 'materials',
  fromMaterialsExpertPage = 'materialsExpert',
  fromExpertsPage = 'experts',
  fromExpertPage = 'expert',
  default = 'default',
}

const configs = {
  postBreadcrumbs: {
    initConfig: [
      {
        pathRegex: /^\/materials\/?$/,
        newState: PostBreadcrumbsState.fromMaterialsPage,
      },
      {
        pathRegex: /^\/experts\/?$/,
        newState: PostBreadcrumbsState.fromExpertsPage,
      },
      {
        pathRegex: /^\/experts\/[1-9][0-9]*\/?$/,
        newState: PostBreadcrumbsState.fromExpertPage,
      },
    ],
    pushConfig: [
      {
        pathRegex: /^\/materials\/?$/,
        newState: PostBreadcrumbsState.fromMaterialsPage,
      },
      {
        pathRegex: /^\/experts\/?$/,
        newState: PostBreadcrumbsState.fromExpertsPage,
      },
    ],
    basedOnPrevConfigType: [
      {
        prevPathRegex: /^\/materials\/?$/,
        currPathRegex: /^\/experts\/[1-9][0-9]*\/?$/,
        newState: PostBreadcrumbsState.fromMaterialsExpertPage,
      },
    ],
    ignoredPaths: [
      /^\/(experts|materials)\/?$/,
      /^\/(experts|posts)\/[1-9][0-9]*\/?$/,
    ],
    defaultState: PostBreadcrumbsState.default,
  },
};

export type BreadcrumbsType = keyof typeof configs;
export const breadcrumbsConfigs: BreadcrumbsContextType = configs;
