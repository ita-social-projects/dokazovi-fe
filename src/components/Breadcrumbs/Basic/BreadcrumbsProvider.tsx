import React, { createContext, useEffect } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { BreadcrumbsType } from './breadcrumbs-config';

interface IProps {
  children: React.ReactNode;
  configs: BreadcrumbsContextType;
}

type PathsConfigType = {
  pathRegex: RegExp;
  newState: string;
};

type BasedOnPrevConfigType = {
  prevPathRegex: RegExp;
  currPathRegex: RegExp;
  newState: string;
};

type IgnoredPathsType = RegExp[];

type BreadcrumbsConfigType = {
  initConfig: PathsConfigType[];
  pushConfig: PathsConfigType[];
  basedOnPrevConfigType: BasedOnPrevConfigType[];
  ignoredPaths: IgnoredPathsType;
  defaultState: string;
};

export type BreadcrumbsContextType = {
  [key: string]: BreadcrumbsConfigType;
};

type BreadcrumbsStateType = {
  breadcrumbsState: string;
  path: string;
};

type BreadcrumbsHistoryStoreType = Map<string, BreadcrumbsStateType>;

const BreadcrumbsContext = createContext<BreadcrumbsContextType>({});

export const useBreadcurmbs = (
  breadcrumbsType: BreadcrumbsType,
  locationKey?: string,
): string | null => {
  const map = getMapFromSessionStorage(breadcrumbsType);
  return map?.get(locationKey ?? '1')?.breadcrumbsState ?? null;
};

export const BreadcrumbsProvider: React.FC<IProps> = ({
  children,
  configs: breadcrumbsContext,
}) => {
  const { key, pathname } = useLocation();
  const { action } = useHistory();

  const updateBreadcrumbsState = (
    sessionKey: string,
    config: BreadcrumbsConfigType,
  ) => {
    let statesMap = getMapFromSessionStorage(sessionKey);

    if (action === 'POP' && !key && statesMap) {
      statesMap.set(
        '1',
        getBreadcrumbsInitState(
          pathname,
          config.initConfig,
          config.defaultState,
        ),
      );
    } else if (!statesMap) {
      statesMap = new Map<string, BreadcrumbsStateType>();
      statesMap.set(
        '1',
        getBreadcrumbsInitState(
          pathname,
          config.initConfig,
          config.defaultState,
        ),
      );
    } else if (action === 'PUSH' && key && statesMap) {
      const prevState = statesMap.get('prevState');
      if (prevState) {
        statesMap.set(
          key,
          getBreacrumbsPushState(
            pathname,
            config.pushConfig,
            config.basedOnPrevConfigType,
            config.ignoredPaths,
            prevState,
            config.defaultState,
          ),
        );
      }
    }

    if (statesMap.has(key ?? '1')) {
      const currBreadcrumbsState = statesMap.get(key ?? '1');
      if (currBreadcrumbsState) {
        statesMap.set('prevState', currBreadcrumbsState);
      }
    }
    setMapToSessionStorage(statesMap, sessionKey);
  };

  useEffect(() => {
    Object.entries(breadcrumbsContext).forEach(([sessionKey, config]) =>
      updateBreadcrumbsState(sessionKey, config),
    );
  }, [pathname]);

  return (
    <BreadcrumbsContext.Provider value={breadcrumbsContext}>
      {children}
    </BreadcrumbsContext.Provider>
  );
};

function getBreadcrumbsInitState(
  pathname: string,
  initConfig: PathsConfigType[],
  defaultState: string,
) {
  const newState =
    initConfig.find((v) => v.pathRegex.test(pathname))?.newState ?? null;

  return {
    path: pathname,
    breadcrumbsState: newState ?? defaultState,
  } as BreadcrumbsStateType;
}

function getBreacrumbsPushState(
  pathname: string,
  pushConfig: PathsConfigType[],
  basedOnPrevConfigType: BasedOnPrevConfigType[],
  excludedPaths: IgnoredPathsType,
  prevState: BreadcrumbsStateType,
  defaultState: string,
) {
  const isExcluded = excludedPaths.some((v) => v.test(pathname));
  const basedOnPrevState = basedOnPrevConfigType.find(
    (v) =>
      v.prevPathRegex.test(prevState.path) && v.currPathRegex.test(pathname),
  );
  const newState =
    basedOnPrevState?.newState ??
    pushConfig.find((v) => v.pathRegex.test(pathname))?.newState ??
    (isExcluded ? prevState.breadcrumbsState : null);

  return {
    path: pathname,
    breadcrumbsState: newState ?? defaultState,
  } as BreadcrumbsStateType;
}

function setMapToSessionStorage(
  breadcrumbsMap: BreadcrumbsHistoryStoreType,
  sessionKey: string,
) {
  const obj = {};
  breadcrumbsMap?.forEach((v, k) => Object.assign(obj, { [k]: v }));
  if (Object.keys(obj).length === 0) {
    return;
  }
  sessionStorage.setItem(sessionKey, JSON.stringify(obj));
}

function getMapFromSessionStorage(sessionKey: string) {
  let map: BreadcrumbsHistoryStoreType;
  try {
    const strMap = sessionStorage.getItem(sessionKey);
    if (!strMap) {
      throw new Error();
    }
    map = new Map<string, BreadcrumbsStateType>(
      Object.entries(JSON.parse(strMap)),
    );
  } catch {
    return null;
  }
  return map;
}
