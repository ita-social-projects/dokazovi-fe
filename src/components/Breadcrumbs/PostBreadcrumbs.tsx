import React from 'react';
import { useLocation } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { langTokens } from '../../locales/localizationInit';
import { PostBreadcrumbsState } from './Basic/breadcrumbs-config';
import { useBreadcurmbs } from './Basic/BreadcrumbsProvider';
import { IOrigin, IPostType } from '../../old/lib/types';
import {
  BasicBreadcrumbs,
  BreadcrumbsLinksType,
} from './Basic/BasicBreadcrumbs';

export interface IProps {
  origins: IOrigin[];
  type: IPostType;
  expert: { id: number; expertName: string };
  materialTitle: string;
}

export const PostBreadcrumbs: React.FC<IProps> = ({
  origins,
  type,
  expert,
  materialTitle,
}) => {
  const { key } = useLocation();
  const { t } = useTranslation();
  const state = useBreadcurmbs('postBreadcrumbs', key);

  const materialsLink = {
    to: '/materials',
    label: t(langTokens.common.materials),
  };
  const expertsLink = {
    to: '/experts',
    label: t(langTokens.common.experts),
  };
  const expertLink = {
    to: `/experts/${expert.id}`,
    label: expert.expertName,
  };
  const originLinks = origins.map((origin) => ({
    to: `/materials?origins=${origin.id}`,
    label: origin.name,
  }));

  const typeLink = { to: `/materials?types=${type.id}`, label: type.name };
  const materialMap = [materialsLink, ...originLinks, typeLink];
  const materialExpertMap = [
    materialsLink,
    expertLink,
    ...originLinks,
    typeLink,
  ];
  const expertsMap = [expertsLink, expertLink, typeLink];
  const expertMap = [expertLink, typeLink];
  const defaultMap = [...originLinks, typeLink];

  let breadcrumbsLinks: BreadcrumbsLinksType;

  switch (state) {
    case PostBreadcrumbsState.fromMaterialsPage: {
      breadcrumbsLinks = materialMap;
      break;
    }
    case PostBreadcrumbsState.fromExpertsPage: {
      breadcrumbsLinks = expertsMap;
      break;
    }
    case PostBreadcrumbsState.fromExpertPage: {
      breadcrumbsLinks = expertMap;
      break;
    }
    case PostBreadcrumbsState.fromMaterialsExpertPage: {
      breadcrumbsLinks = materialExpertMap;
      break;
    }
    default: {
      breadcrumbsLinks = defaultMap;
    }
  }

  return (
    <BasicBreadcrumbs
      breadcrumbsLinks={breadcrumbsLinks}
      title={materialTitle}
    />
  );
};
