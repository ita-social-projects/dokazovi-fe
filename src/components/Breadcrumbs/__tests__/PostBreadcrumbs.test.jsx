/* eslint-disable @typescript-eslint/no-unsafe-call */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { PostBreadcrumbs } from '../PostBreadcrumbs';
import { useBreadcurmbs } from '../Basic/BreadcrumbsProvider';

jest.mock('../Basic/BreadcrumbsProvider', () => ({
  useBreadcurmbs: jest.fn(),
}));

const postBreadcrumbsMocks = {
  origins: [{ id: 2, name: 'Медитека', parameter: null }],
  type: { id: 1, name: 'Стаття' },
  expert: { id: 1, expertName: 'Єва Євенко' },
  materialTitle: 'Як підготуватися до вакцинації',
};

describe('PostBreadcrumbs tests', () => {
  it('should PostBreadcrumbs render', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <PostBreadcrumbs
          origins={postBreadcrumbsMocks.origins}
          type={postBreadcrumbsMocks.type}
          expert={postBreadcrumbsMocks.expert}
          materialTitle={postBreadcrumbsMocks.materialTitle}
        />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should materials breadcrumbs be visible', () => {
    useBreadcurmbs.mockImplementation(() => 'materials');

    render(
      <MemoryRouter>
        <PostBreadcrumbs
          origins={postBreadcrumbsMocks.origins}
          type={postBreadcrumbsMocks.type}
          expert={postBreadcrumbsMocks.expert}
          materialTitle={postBreadcrumbsMocks.materialTitle}
        />
      </MemoryRouter>,
    );

    expect(screen.getAllByTestId('link')[0]).toHaveAttribute(
      'href',
      '/materials',
    );

    expect(screen.getAllByTestId('link')[1]).toHaveAttribute(
      'href',
      '/materials?origins=2',
    );

    expect(screen.getAllByTestId('link')[2]).toHaveAttribute(
      'href',
      '/materials?types=1',
    );
  });

  it('should materialsExpert breadcrumbs be visible', () => {
    useBreadcurmbs.mockImplementation(() => 'materialsExpert');

    render(
      <MemoryRouter>
        <PostBreadcrumbs
          origins={postBreadcrumbsMocks.origins}
          type={postBreadcrumbsMocks.type}
          expert={postBreadcrumbsMocks.expert}
          materialTitle={postBreadcrumbsMocks.materialTitle}
        />
      </MemoryRouter>,
    );

    expect(screen.getAllByTestId('link')[0]).toHaveAttribute(
      'href',
      '/materials',
    );

    expect(screen.getAllByTestId('link')[1]).toHaveAttribute(
      'href',
      '/experts/1',
    );
  });

  it('should experts breadcrumbs be visible', () => {
    useBreadcurmbs.mockImplementation(() => 'experts');

    render(
      <MemoryRouter>
        <PostBreadcrumbs
          origins={postBreadcrumbsMocks.origins}
          type={postBreadcrumbsMocks.type}
          expert={postBreadcrumbsMocks.expert}
          materialTitle={postBreadcrumbsMocks.materialTitle}
        />
      </MemoryRouter>,
    );

    expect(screen.getAllByTestId('link')[0]).toHaveAttribute(
      'href',
      '/experts',
    );

    expect(screen.getAllByTestId('link')[1]).toHaveAttribute(
      'href',
      '/experts/1',
    );

    expect(screen.getAllByTestId('link')[2]).toHaveAttribute(
      'href',
      '/materials?types=1',
    );
  });

  it('should expert breadcrumbs be visible', () => {
    useBreadcurmbs.mockImplementation(() => 'expert');

    render(
      <MemoryRouter>
        <PostBreadcrumbs
          origins={postBreadcrumbsMocks.origins}
          type={postBreadcrumbsMocks.type}
          expert={postBreadcrumbsMocks.expert}
          materialTitle={postBreadcrumbsMocks.materialTitle}
        />
      </MemoryRouter>,
    );

    expect(screen.getAllByTestId('link')[0]).toHaveAttribute(
      'href',
      '/experts/1',
    );

    expect(screen.getAllByTestId('link')[1]).toHaveAttribute(
      'href',
      '/materials?types=1',
    );
  });
});
