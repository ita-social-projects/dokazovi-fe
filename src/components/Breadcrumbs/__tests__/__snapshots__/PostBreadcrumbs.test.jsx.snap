// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostBreadcrumbs tests should PostBreadcrumbs render 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
    class="MuiTypography-root MuiBreadcrumbs-root BasicBreadcrumbs-breadcrumbs-1 MuiTypography-body1 MuiTypography-colorTextSecondary"
  >
    <ol
      class="MuiBreadcrumbs-ol"
    >
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          class="BasicBreadcrumbs-link-2"
          href="/"
        >
          Головна
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          class="BasicBreadcrumbs-link-2"
          data-testid="link"
          href="/materials?origins=2"
        >
          Медитека
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <a
          class="BasicBreadcrumbs-link-2"
          data-testid="link"
          href="/materials?types=1"
        >
          Стаття
        </a>
      </li>
      <li
        aria-hidden="true"
        class="MuiBreadcrumbs-separator"
      >
        /
      </li>
      <li
        class="MuiBreadcrumbs-li"
      >
        <p
          class="MuiTypography-root BasicBreadcrumbs-postTitle-3 MuiTypography-body1"
        >
          Як підготуватися до вакцинації
        </p>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;
